### PR TITLE
remove DEBUG_BUILD_OPTIMIZATION_ONE

### LINKED ISSUE #283: Bad assembly with SimpleArray caster
In the [SimpleArrayCaster.hpp](https://github.com/solvcon/modmesh/pull/266) file, a custom caster for **SimpleArray** was defined to cast from **SimpleArrayPlex** to **SimpleArray**. By default, the debug build uses the **-O0** flag and the binary is incorrect. The compiler ignores the custom caster and opts for the generic caster instead. Setting the optimization level to **-O1** or higher leads to a correct binary and the custom caster works correctly.

This anomaly was identified on Windows (tested with MSVC 2022) and macOS (tested with LLVM 15 and GCC 13 toolchains). Notably, Ubuntu 22.04 with GCC 11 did not exhibit this issue.

The issue was initially observed in [PR#266](https://github.com/solvcon/modmesh/pull/266), and a concise investigation report is available [here](https://github.com/solvcon/modmesh/pull/266#issuecomment-1946362573).

As a temporary solution, a workaround involves adding a flag, **DEBUG_BUILD_OPTIMIZATION_ONE**, in CMake to ensure successful CI passes on Windows and macOS by applying **-O1** for the debug build. By default, this flag should be set to **OFF** to reproduce the issue. Developers are advised to apply this CMake flag to ensure correct behavior in their local environments, aligning with CI expectations.

### PR DESCRIPTION
**Pull request by tigercosmos**
To address #283, two steps should be applied.
- put the `SimpleArrayCaster.hpp` into `array_common.hpp` (check more at the comments in the code)
- refactored `wrap_modifiers()`, which leads some type issues for pybind11 when we introduce customized caster for the array.

close #283

### CODE REVIEW

File: .github/workflows/devbuild.yml
Comment (tigercosmos) [inline .github/workflows/devbuild.yml:491]: removed all `DEBUG_BUILD_OPTIMIZATION_ONE` in the CI

File: CMakeLists.txt
Comment (tigercosmos) [inline CMakeLists.txt:101]: Deleting these is a bit regrettable

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/array_common.hpp:42]: please see the comments.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: it's interesting that we need to write in this way to make pybind11 work correctly.

File: cpp/modmesh/python/common.hpp
Comment (tigercosmos) [inline cpp/modmesh/python/common.hpp]: use modern C++

**Comment by tigercosmos** [comment]
@yungyuc ready for review.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: This is interesting, and I used some similar workarounds before with boost.python.

Does using lambda resolve the bad assembly?  If yes, please add comments here to explain to future maintainers.  Also, try to keep `wrap_modifiers()`.

File: cpp/modmesh/python/common.hpp
Comment (yungyuc) [reply cpp/modmesh/python/common.hpp]: If this does nothing to the resolution of the bad assembly, please avoid the refactoring change in this **changeset**.

You may use another refactoring PR.

**Review by yungyuc — COMMENTED** [review]
Thanks for the update.

- [x] Discuss lambda.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: sure, kept `wrap_modifiers` and used lambda inside.

File: cpp/modmesh/python/common.hpp
Comment (tigercosmos) [reply cpp/modmesh/python/common.hpp]: sure

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: 1. The lambda isn't responsible for the problematic assembly; instead, putting `SimpleArrayCaster.hpp` into `array_common.hpp` addresses it (I don't know why this works, to be honested)
2. Upon integrating `SimpleArrayCaster.hpp` into wrap_`SimpleArray.cpp`, a lambda becomes necessary to rectify the erroneous implicit casting in Pybind11. Without the lambda, Pybind11 fails to build due to certain type issues, albeit not explored in depth.

**Comment by tigercosmos** [comment]
@yungyuc  fixed the issues above.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: Please add the issue #283 in the comment for easy reference.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp:157]: I tweaked the comment with the link to the issue #283 added.

**Review by yungyuc — APPROVED** [review]
LGTM.  Waiting for CI and I will merge.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: Oh this is even more interesting.  There is definitely something we missed or misunderstood in the language and/or the compiler.  But pragmatism beats purity.  Let's get this to work and keep the experience.  Maybe someday the Lord will shed lights on this matter.

This is good finding and description, and will be a good reference in the future.

**Comment by tigercosmos** [comment]
thanks @yungyuc

**Pull request merged**
